### PR TITLE
add magic method XCom docstrings D105

### PR DESCRIPTION
**Pull request by okirialbert**
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by okirialbert** [comment]
Add XCom magic method docstrings #37523

**Review by vincbeck — APPROVED** [review]

File: airflow/models/xcom.py [hunk 1]
```diff
@@ -131,6 +131,12 @@ def init_on_load(self):
         self.value = self.orm_deserialize_value()
 
     def __repr__(self):
+        """Return the XCom properties.
```
Comment (Taragolis) [inline airflow/models/xcom.py]: `__repr__` is stand for string representation
`__str__` is also sting representation, however it usual use when you need to convert object to string (every object in Python in theory should be convertable to sting), e.g. `str(object)` or `f"Foo Bar {object}"`

https://docs.python.org/3/reference/datamodel.html#object.__repr__

File: airflow/models/xcom.py [hunk 2]
```diff
@@ -785,19 +801,26 @@ def __getstate__(self) -> Any:
...
 
     def __setstate__(self, state: Any) -> None:
+        """Construct a query by retrieving the statement from the `state` and assign it to the `self._query`class property."""
         statement, self._len = state
         self._query = Query(XCom.value).from_statement(text(statement))
 
     def __len__(self):
+        """Declare a `self._len` class property that assigns the number of rows the query returns using `query.count()` if the `self._len` property is `None`.
+        
+        Otherwise return `self._len`
```
Comment (Taragolis) [inline airflow/models/xcom.py]: Some more specific better here, `__len__` just about length of the object. Better avoid to detail explanation into the docstring. Implementation could be changed, and docstring transformed into the "liar"

You could ask a round in `#new-contributors` or `#contributors` slack channel (join link if you not there https://s.apache.org/airflow-slack)  if you would like what kind of the object it calculate length

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -785,19 +801,26 @@ def __getstate__(self) -> Any:
...
+        """
         if self._len is None:
             with self._get_bound_query() as query:
                 self._len = query.count()
         return self._len
 
     def __iter__(self):
+        """Return a `_LazyXComAccessIterator` class instance to manage query sessions."""
```
Comment (Taragolis) [inline airflow/models/xcom.py]: Better replace by Iterate over the ..., and try to avoid mention private/internal classes because them not part of the public API

**Review by potiuk — CHANGES_REQUESTED** [review]
I find it odd that we need to describe the dunder methods (__str__ ) - if that's the case then the rule  should not be enabled. 

It's simply wrong to describe what those methods do. Generally those descriptions, describe what is implemented in them (not what they do) - which is duplication and we should never try to attempt to describe in the docstring in plain English what is internal and plain visible  in code - we should rather explain what is the intention of the methods (and we already know that because we know what the dunder methods in question do (`__repr__`, `__eq_`) . Repeating and explaining the algorithm used internally is really bad - because it's almost 100% certain that when we change the implementation (and for example add one more field in `__repr__` ) we will forget to also update description and it will be just lying to us.

That makes value of those docstrings negative, because they will be misleading the user of those methods

**Comment by okirialbert** [comment]
I was striving for better readability by including detailed function descriptions but now I see how it can be misleading in the event of future changes.

File: airflow/models/xcom.py (hunk 3 above)
Comment (okirialbert) [reply airflow/models/xcom.py]: Noted. I'll make the changes.

**Comment by uranusjr** [comment]
Magic methods already have defined semantics and documenting them is entirely missing the point. If a tool requires you to do it, it is not a good tool.

**Comment by potiuk** [comment]
> Magic methods already have defined semantics and documenting them is entirely missing the point. If a tool requires you to do it, it is not a good tool.

Yep. Agree with @uranusjr - we shoudl NOT be documenting those (`dunder`) methods at all.

**Comment by Taragolis** [comment]
My 2c, there is always depends on. E.g. why on earth we required to create `__str__` and `__repr__` for the classes:
- is this just for beatify? 
- Or it has some specific cases, e.g. https://github.com/apache/airflow/blob/c0b849ad2b3f7015f7cb2a45aefd1fa3828bda31/airflow/models/xcom_arg.py#L277-L286

`__repr__` might have a specific meaning for the serialisation, e.g.

```python
class Foo:
    ...

class Bar:
    def __repr__(self):
        return "BarSentinel"
```

e.g. `Foo` will return a new value each time it created `<__main__.Foo object at 0x1011e5610>` or in another time it would be `<__main__.Foo object at 0x101065610>`. So if this object might use into the serialisation, then it will be the reason why we overwrite SerializedDag every time, and simple "String representation for deterministic output" might help understand why it required to set this method.

The problem here that for some methods you have to define, e.g. if you subclassing of `collections.abc`, and that is pretty clear why you have to define and docstring in this case redundant

I have not clear vision, should we define it for all classes (rule enable) or it should remain on conscience and responsibility of contributor/reviewers (rule disable).

And there other question, do we have an idea why we originally (3 ago) select this rules https://github.com/apache/airflow/issues/10742 . For example Personally I thought it is redundant to also enable `D107 | Missing docstring in __init__`, so maybe better review and discuss this rules again and remove redundant one?
